src/os/command/feature/featurefillopacitycmd.js

```diff
@@ -0,0 +1,80 @@
+goog.provide('os.command.FeatureFillOpacity');
+
+goog.require('os.command.AbstractFeatureStyle');
+goog.require('os.events.PropertyChangeEvent');
+goog.require('os.metrics');
+goog.require('os.ui');
+
+
+
+/**
+ * Changes the fill opacity of a feature
+ * @extends {os.command.AbstractFeatureStyle}
+ * @param {string} layerId
+ * @param {string} featureId
+ * @param {number} opacity
+ * @param {number=} opt_oldOpacity
+ * @constructor
+ */
+os.command.FeatureFillOpacity = function(layerId, featureId, opacity, opt_oldOpacity) {
+  os.command.FeatureFillOpacity.base(this, 'constructor', layerId, featureId, opacity, opt_oldOpacity);
+  this.title = 'Change Feature Fill Opacity';
+  this.metricKey = os.metrics.Layer.FEATURE_FILL_OPACITY;
+
+  this.value = opacity;
+};
+goog.inherits(os.command.FeatureFillOpacity, os.command.AbstractFeatureStyle);
+
+
+/**
+ * @type {number}
+ * @const
+ */
+os.command.FeatureFillOpacity.DEFAULT_OPACITY = 1;
+
+
+/**
+ * @inheritDoc
+ */
+os.command.FeatureFillOpacity.prototype.getOldValue = function() {
+  var feature = /** @type {ol.Feature} */ (this.getFeature());
+  var config = /** @type {Array<Object>|Object|undefined} */ (this.getFeatureConfigs(feature));
+  if (goog.isArray(config)) {
+    config = config[0];
+  }
+
+  return config ? os.style.getConfigOpacityColor(config) : os.command.FeatureFillOpacity.DEFAULT_OPACITY;
+};
+
+
+/**
+ * @inheritDoc
+ */
+os.command.FeatureFillOpacity.prototype.applyValue = function(configs, value) {
+  var color;
+  for (var i = 0; i < configs.length; i++) {
+    color = os.style.getConfigColor(configs[i], true, os.style.StyleField.FILL);
+    color[3] = value;
+    os.style.setConfigColor(configs[i], color, [os.style.StyleField.FILL]);
+  }
+
+  os.command.FeatureFillOpacity.base(this, 'applyValue', configs, value);
+};
+
+
+/**
+ * @inheritDoc
+ */
+os.command.FeatureFillOpacity.prototype.finish = function(configs) {
+  // dispatch the fill opacity change event on the source for the histogram
+  var feature = /** @type {ol.Feature} */ (this.getFeature());
+  var source = /** @type {plugin.file.kml.KMLSource} */ (os.feature.getSource(feature));
```

Comment:
This cast is referencing the KML plugin in core code, and is likely not making a correct typecast assumption about the type of source it's getting in all cases. There should be a better way to fire change events on the features themselves that then get handled by their nodes to do KML-specific stuff. It may also be more appropriate for this change to be notified on the source. @schmidtk would know better than me.

Reply:
The command shouldn't have to use the source at all. It should fire a change event on the feature, which should be handled by the source/node. In this case, you should fire an event on the feature and handle it in `plugin.file.kml.ui.KMLNode#onFeatureChange`. The handler should respond by firing a new event from the node.

Reply:
Admittedly this seems to be copied from my other implementations of feature commands ... so this would probably need to be fixed for all of those commands

Reply:
This should now be improved; the plugin references have been moved to just plugin code. I will squash commits at the end to add the breaking change message.